### PR TITLE
hook up mslk's to_nvfp4 kernel to torchao's inference nvfp4 workflows

### PR DESCRIPTION
**Pull request by vkuzo**
Summary:

Changes NVFP4 inference triton kernel to use `mslk` instead of the one checked in to torchao.  Note that `mslk` (an optional dependency) is now required for the default usage of `NVFP4DynamicActivationNVFP4WeightConfig`.

* microbenchmark speedups - large
  * for example dynamic + MKN 4096 speedup increases from 0.90x to 1.27x.  See full table below for more.
* e2e speedups on flux-1.schnell - medium
  * 1.26x -> 1.30x at bsz 1
  * 1.32x -> 1.37x at bsz 4
* e2e accuracy - no change

We can delete the torchao's nvfp4 kernel in a future PR, to keep this one small.

Currently `torchao` defines the nvfp4 global scale as `amax / (448 * 6)`, and mslk (and flashinfer) define it as `(448 * 6) / amax`. In a future PR we should consider moving torchao to the mslk definition.  Saving that for a future PR as that will be a BC-breaking change which will need careful testing of existing checkpoints.

Test Plan:

microbenchmark sweep before

```
> python benchmarks/float8/float8_inference_roofline.py --recipe_name nvfp4 --enable_fusion_modeling True --skip_printing_detailed_metrics True

0   1024   1024   1024                      1.00            0.28
1   2048   2048   2048                      2.36            0.52
2   4096   4096   4096                      2.89            0.90
3   8192   8192   8192                      3.32            1.41
4  16384  16384  16384                      3.62            2.14

> python benchmarks/float8/float8_inference_roofline.py --recipe_name nvfp4_static --enable_fusion_modeling True --skip_printing_detailed_metrics True

   fwd_M  fwd_K  fwd_N  r_fp8_gemm_and_ovhd_spdp  b_fp8_e2e_spdp
0   1024   1024   1024                      1.00            0.34
1   2048   2048   2048                      2.74            0.64
2   4096   4096   4096                      3.42            1.06
3   8192   8192   8192                      3.67            1.58
4  16384  16384  16384                      3.82            2.31
```

and after

```
> python benchmarks/float8/float8_inference_roofline.py --recipe_name nvfp4 --enable_fusion_modeling True --skip_printing_detailed_metrics True

   fwd_M  fwd_K  fwd_N  r_fp8_gemm_and_ovhd_spdp  b_fp8_e2e_spdp
0   1024   1024   1024                      1.00            0.39
1   2048   2048   2048                      2.36            0.68
2   4096   4096   4096                      2.89            1.27
3   8192   8192   8192                      3.32            1.93
4  16384  16384  16384                      3.62            2.73

> python benchmarks/float8/float8_inference_roofline.py --recipe_name nvfp4_static --enable_fusion_modeling True --skip_printing_detailed_metrics True

   fwd_M  fwd_K  fwd_N  r_fp8_gemm_and_ovhd_spdp  b_fp8_e2e_spdp
0   1024   1024   1024                      1.00            0.48
1   2048   2048   2048                      2.74            0.88
2   4096   4096   4096                      3.42            1.62
3   8192   8192   8192                      3.67            2.27
4  16384  16384  16384                      3.82            2.98

```

e2e model accuracy

```
./benchmarks/quantization/eval_accuracy_and_perf_of_flux.sh
```

### CODE REVIEW

**Comment by vkuzo** [comment]
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.14.0) (oldest at bottom):
* #4078
* #4077
* __->__ #4031

File: test/prototype/mx_formats/test_nvfp4_tensor.py [hunk 1]
```diff
@@ -413,6 +418,51 @@ def test_triton_nvfp4_quantize_equivalence(M, N, use_per_tensor_scale, dtype):
...
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="CUDA not available")
+@pytest.mark.parametrize(
+    "M", [128, 256, 512, 1024, 100, 200, 384], ids=lambda m: f"M{m}"
+)
+@pytest.mark.parametrize("N", [64, 128, 256, 512, 32, 96, 160], ids=lambda n: f"N{n}")
+@pytest.mark.skipif(
+    not is_sm_at_least_100(), reason="requires sm100+ for nvfp4 triton kernel"
+)
+@torch.no_grad()
+def test_mslk_nvfp4_numerics(M, N):
```
Comment (vkuzo) [inline test/prototype/mx_formats/test_nvfp4_tensor.py]: need to combine this with test_triton_nvfp4_quantize_equivalence before land

Comment (vkuzo) [reply test/prototype/mx_formats/test_nvfp4_tensor.py]: the other test is properly gated for sm100 as well

**Comment by vkuzo** [comment]
still seeing e2e regressions on diffusion models, need to chase it down

**Comment by drisspg** [comment]
@claude review this pr, determine if there are any flaws in benchmarking methodology

**Comment by vkuzo** [comment]
update: the e2e slowdowns I saw were in models where the gpu was not saturated.  Things look good if we turn on cuda graphs.

Going to solidify nvfp4 e2e benchmarking, then rebase this PR on top so we clearly see the e2e wins.

**Comment by drisspg** [comment]
Are there benchmark runs that report memory throughput?

**Comment by vkuzo** [comment]
> Are there benchmark runs that report memory throughput?

we don't have it in torchao right now, but I validated with https://github.com/sayakpaul/diffusers-blackwell-quants that peak GPU memory does not change with this PR

File: torchao/prototype/mx_formats/kernels.py [hunk 2]
```diff
@@ -1387,3 +1387,72 @@ def mxfp8_quantize_cuda(
         raise NotImplementedError(
             "`mxfp8_quantize_cuda` needs (1) torch 2.8+ and (2) torchao built from source on a machine with CUDA capability 10.0+. Please see https://github.com/pytorch/ao/issues/2932 for more details."
         )
+
+
+try:
```
Comment (drisspg) [inline torchao/prototype/mx_formats/kernels.py]: nit: I think importlib is a better pattern here

File: torchao/prototype/mx_formats/kernels.py [hunk 3]
```diff
@@ -1387,3 +1387,72 @@ def mxfp8_quantize_cuda(
...
+    """Quantize a tensor to NVFP4 using the MSLK triton kernel.
+
+    Args:
+        x: Input tensor to quantize.
+        per_tensor_scale: Per-tensor scale (TorchAO convention: amax / (F8E4M3_MAX * F4_E2M1_MAX)).
+
+    Returns:
+        Tuple of (blockwise_scales, quantized_data_uint8) matching TorchAO's convention.
+    """
+    mslk_global_scale = 1.0 / per_tensor_scale
```
Comment (drisspg) [inline torchao/prototype/mx_formats/kernels.py]: other nit, I like recpirocal I have a memory of nikita trying claiming this was better but cant remember now

File: torchao/prototype/mx_formats/kernels.py (hunk 2 above)
Comment (vkuzo) [reply torchao/prototype/mx_formats/kernels.py]: link to what you mean?

File: torchao/prototype/mx_formats/kernels.py (hunk 3 above)
Comment (vkuzo) [reply torchao/prototype/mx_formats/kernels.py]: you mean `per_tensor_scale.reciprocal()`?

Comment (drisspg) [reply torchao/prototype/mx_formats/kernels.py]: yea

File: torchao/prototype/mx_formats/kernels.py (hunk 2 above)
Comment (drisspg) [reply torchao/prototype/mx_formats/kernels.py]: Actually this is fine; https://github.com/pytorch/pytorch/blob/004e7e6e96992aedab1949e20aadd6b8837c1489/torch/_inductor/kernel/flex/flex_flash_attention.py#L71-L74

In theory, you shouldn't have to try to accept this code, but I remember running into a bug where I needed to. In the end, these are basically morally equivalent. This is fine
